File: ch03/entropy.go
package main

import (
	"fmt"
	"math"
)

func calculateEntropy(dataSet [][]string) float64 {
	labelCounts := make(map[string]int)
	for _, vector := range dataSet {
		label := vector[len(vector)-1]
		labelCounts[label]++
	}

	var entropy float64
	totalEntries := float64(len(dataSet))
	for _, count := range labelCounts {
		p := float64(count) / totalEntries
		entropy -= p * math.Log2(p)
	}
	return entropy
}

func splitDataSet(dataSet [][]string, featureIndex int, value string) [][]string {
	var subset [][]string
	for _, vec := range dataSet {
		if vec[featureIndex] == value {
			reducedVec := append(vec[:featureIndex], vec[featureIndex+1:]...)
			subset = append(subset, reducedVec)
		}
	}
	return subset
}

func chooseBestFeature(dataSet: [][]string) int {
	numFeatures := len(dataSet[0]) - 1
	baseEntropy := calculateEntropy(dataSet)
	bestInfoGain := 0.0
	bestFeature := -1

	for i := 0; i < numFeatures; i++ {
		featureValues := make(map[string]bool)
		for _, vector := range dataSet {
			featureValues[vector[i]] = true
		}
		var newEntropy float64
		for value := range featureValues {
			subset := splitDataSet(dataSet, i, value)
			p := float64(len(subset)) / float64(len(dataSet))
			newEntropy += p * calculateEntropy(subset)
		}
		infoGain := baseEntropy - newEntropy
		if infoGain > bestInfoGain {
			bestInfoGain = infoGain
			bestFeature = i
		}
	}
	return bestFeature
}

func createTree(dataSet [][]string, featureLabels []string) map[string]interface{} {
	classList := make([]string, len(dataSet))
	for i, vector := range dataSet {
		classList[i] = vector[len(vector)-1]
	}

	allSame := true
	for i := 1; i < len(classList); i++ {
		if classList[i] != classList[0] {
			allSame = false
			break
		}
	}
	if allSame {
		return map[string]interface{}{"label": classList[0]}
	}

	if len(dataSet[0]) == 1 {
		labelCounts := make(map[string]int)
		for _, label := range classList {
			labelCounts[label]++
		}
		var majorityLabel string
		maxCount := 0
		for label, count := range labelCounts {
			if count > maxCount {
				maxCount = count
				majorityLabel = label
			}
		}
		return map[string]interface{}{"label": majorityLabel}
	}
	
	bestFeature := chooseBestFeature(dataSet)
	bestFeatureLabel := featureLabels[bestFeature]
	tree := make(map[string]interface{})
	tree[bestFeatureLabel] = make(map[string]interface{})
	featureValues := make(map[string]bool)
	for _, vector := range dataSet {
		featureValues[vector[bestFeature]] = true
	}

	for value := range featureValues {
		subLabels := append([]string{}, featureLabels...)
		subLabels = append(subLabels[:bestFeature], subLabels[bestFeature+1:]...)
		subDataSet := splitDataSet(dataSet, bestFeature, value)
		tree[bestFeatureLabel].(map[string]interface{})[value] = createTree(subDataSet, subLabels)
	}
	return tree
}

func classify(tree map[string]interface{}, featureLabels []string, sample []string) string {
	rootFeature := ""
	for k := range tree {
		rootFeature = k
		break
	}
	featureIndex := -1
	for i, label := range featureLabels {
		if label == rootFeature {
			featureIndex = i
			break
		}
	}
	subTree := tree[rootFeature].(map[string]interface{})
	sampleValue := sample[featureIndex]
	var result string
	if nextNode, ok := subTree[sampleValue]; ok {
		switch nextNode.(type) {
		case map[string]interface{}:
			result = classify(nextNode.(map[string]interface{}), append(featureLabels[:featureIndex], featureLabels[featureIndex+1:]...), append(sample[:featureIndex], sample[featureIndex+1:]...))
		case string:
			result = nextNode.(string)
		}
	} else {
		result = "Unknown"
	}
	return result
}

File: ch03/entropy.py
from math import log


def entropy(dataSet):
    print("Calculating entropy for the dataset...")

    labels = [row[-1] for row in dataSet]
    counts = {}
    for label in labels:
        counts[label] = counts.get(label, 0) + 1
    entropy = 0.0
    total = len(dataSet)
    for c in counts.values():
        prob = c / total
        entropy -= prob * log(prob, 2)

    return entropy


def splitDataSet(dataSet, axis, value):
    print(f"Splitting dataset on axis {axis} for value {value}...")
    return [row[:axis] + row[axis + 1 :] for row in dataSet if row[axis] == value]


def chooseBestFeatureToSplit(dataSet):
    print("Choosing the best feature to split on...")
    numFeatures = len(dataSet[0]) - 1
    baseEntropy = entropy(dataSet)
    bestGain = 0.0
    bestFeature = -1

    for i in range(numFeatures):
        values = set(row[i] for row in dataSet)
        newEntropy = 0.0
        for v in values:
            subset = splitDataSet(dataSet, i, v)
            weight = len(subset) / len(dataSet)
            newEntropy += weight * entropy(subset)
        gain = baseEntropy - newEntropy
        if gain > bestGain:
            bestGain = gain
            bestFeature = i

    return bestFeature

def createTree(dataSet, featureNames):
    print("Creating decision tree...")
    labels = [row[-1] for row in dataSet]
    if labels.count(labels[0]) == len(labels):
        return labels[0]
    if len(dataSet[0]) == 1:
        return max(set(labels), key=labels.count)

    bestFeature = chooseBestFeatureToSplit(dataSet)
    bestFeatureName = featureNames[bestFeature]
    tree = {bestFeatureName: {}}
    del featureNames[bestFeature]

    values = set(row[bestFeature] for row in dataSet)
    for v in values:
        subset = splitDataSet(dataSet, bestFeature, v)
        subtree = createTree(subset, featureNames[:])
        tree[bestFeatureName][v] = subtree

    return tree